### ISSUE TITLE
TFLite C++ API outputs the same results all the time

### ISSUE THREAD

**Issue by flamxi** [tensorflow/tensorflow#43788]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): **Yes**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): **Linux Ubuntu 18.04**
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: **HDK865 from QC**
- TensorFlow installed from (source or binary): **Python binary, C++ API from source**
- TensorFlow version (use command below): **2.3**
- Python version: **3.6**
- Bazel version (if compiling from source): **Build label: 3.5.0**
- GCC/Compiler version (if compiling from source): **7.5.0** 
- CUDA/cuDNN version:
- GPU model and memory:

I have an MLP model trained with **Keras** and **Tensorflow 2.3**, I train and predict correctly with that model as .tflite using Python API, but when I try to predict using C++ API output is always the same 1/model_classes, all the time.

I feed the model: 
```c++ 
void fillInputTenso(std::vector<float> input) {
        auto inputTensor = this->mInterpreter->typed_tensor<float>(this->mInterpreter->inputs()[0]);
        for(int i = 0; i < input.size(); i++) {
            inputTensor[i] = input[i];
        }
```
After Invoke I read the output: 
```c++ 
 void proceedOutput(void) {
        auto output = this->mInterpreter->typed_tensor<float>(this->mInterpreter->outputs()[0]);
        for(int i=0; i < this->mConfig.labels.size(); i++) {
            LOGI(" confidence = %f.", output[i]);
        }
    }
```
I also tried this from the C++ API but the results are the same: 
```c++
auto input = mInterpreter->typed_input_tensor<float>(0) 
```
**It doesn't matter what input I provide to model, its completely the same output always.
This model works completely fine with Python API.**

Model architecture:
```python
keras.Sequential([
            keras.layers.Flatten(input_shape=(self.dataset.features_num(),)),
            keras.layers.Dense(128, activation='relu'),
            keras.layers.Dense(nr_classes, activation='softmax')
        ])
```
Note: I added this node to this model since the accuracy was dropped drastically when converted to tflite:
```python
import tensorflow_model_optimization as tfmot
            quantize_model = tfmot.quantization.keras.quantize_model
            # q_aware stands for for quantization aware.
            q_aware_model = quantize_model(self.__model)
            # `quantize_model` requires a recompile.
            q_aware_model.compile(optimizer='adam',
                                  loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
                                  metrics=['accuracy'])
            self.__model = q_aware_model
```

**Comment by srjoglekar246** [comment]
Can you provide some more details about the process you are employing to BUILD & run with the C++ API?

Also, a snippet of the code where you allocate tensors & invoke the C++ interpreter would help.

**Comment by flamxi** [comment]
This is my BUILD file.

![image](https://user-images.githubusercontent.com/30228544/95462317-a2adf600-0977-11eb-82ca-a416d11b6417.png)

This is the part when I init Interpreter and AllocateTensors:

```c++
 tflite::StderrReporter error_reporter;
        auto model = tflite::FlatBufferModel::BuildFromFile(model_path, &error_reporter);
        tflite::ops::builtin::BuiltinOpResolver resolver;
        
        if(kTfLiteOk != tflite::InterpreterBuilder(*(model), resolver)(&this->mInterpreter)) {
            LOGE("InterpreterBuilder failed. Exiting!");
            exit(-1);
        }
        
        this->mInterpreter->SetNumThreads(threads);
        // this->mInterpreter->UseNNAPI(true);


        if(kTfLiteOk != this->mInterpreter->AllocateTensors()) {
            LOGE("interpreter->AllocateTensor failed. Exiting!");
            exit(-1);
        }
```

I invoke:

```c++
if(kTfLiteOk != mInterpreter->Invoke()) {
            LOGE("mInterpreter->Invoke() failed. Exiting!");
            exit(-1);
        }

```

**Comment by srjoglekar246** [comment]
Hmm. At first glance, I can't find anything wrong with your code. I assume you are persisting the `model` data returned here, until the lifetime of the interpreter:
```
auto model = tflite::FlatBufferModel::BuildFromFile(model_path, &error_reporter);
```

Apart from that, it's difficult to say what could be going wrong without looking into your model.

I would suggest using a float model first to check everything works in that case, but it shouldn't happen that the Python bindings work but the native runtime fails.

**Comment by flamxi** [comment]
Hey dude, in fact `model` it had its lifetime only inside that function scope, I changed it as you mentioned now exists for the whole interpreter lifetime and all works smoothly. You sir, are a life saver :). 
May I ask why its needed for the model data to exists for the whole time of interpreter, and why isn't any segmentation fault or smh if it is needed!

**Comment by srjoglekar246** [comment]
So TFLite, to save memory, does not copy constants into a different memory location. They are read from a memory-mapped version of the model data (`model` in this case). So if you destroy the model file, the Interpreter looks at garbage and does its computations. Sometimes this leads to segfaults (if the data for say `shape` is garbage integers), but that may not always happen in (un)fortunate cases like yours :-).
